src/Umbraco.Core/Services/IdkMap.cs

```diff
@@ -156,16 +163,72 @@ public void ClearCache(Guid key)
             try
             {
                 _locker.EnterWriteLock();
-                int id;
-                if (_key2Id.TryGetValue(key, out id) == false) return;
-                _id2Key.Remove(id);
-                _key2Id.Remove(key);
+                var match = _key2Id.Keys.SingleOrDefault(x => x.Key == key);
+                if (match == null) return;
+                var id = _key2Id[match];
+                var reversedCompositeKey = new Id2KeyCompositeKey {Id = id, UmbracoObjectType = match.UmbracoObjectType};
+                _id2Key.Remove(reversedCompositeKey);
+                _key2Id.Remove(match);
             }
             finally
             {
                 if (_locker.IsWriteLockHeld)
                     _locker.ExitWriteLock();
             }
         }
+
+        internal class Id2KeyCompositeKey
+        {
+            public int Id { get; set; }
```

Comment:
Yep! Actually thought about the struct thing myself this morning. The immutable thing is a good point - keys should not be immutable at all. I'll fix it up :)